File: projects/phase-02-computer-science-foundations/cs50x/week-01-c/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long card_number = get_long_long("Number: ");
    long long card_number_copy = card_number;

    int count = 0;

    while (card_number_copy > 0)
    {
        card_number_copy /= 10;
        count++;
    }

    card_number_copy = card_number;

    int sum1 = 0;
    int sum2 = 0;

    while (card_number_copy > 0)
    {
        sum2 += card_number_copy % 10;

        card_number_copy /= 10;

        if (card_number_copy == 0)
        {
            break;
        }

        int number = card_number_copy % 10;
        number *= 2;

        if (number > 9)
        {
            sum1 += number % 10;
            sum1 += number / 10;
        }

        else
        {
            sum1 += number;
        }

        card_number_copy /= 10;
    }

    int total = sum1 + sum2;

    card_number_copy = card_number;

    while (card_number_copy >= 100)
    {
        card_number_copy /= 10;
    }

    if (total % 10 != 0)
    {
        printf("INVALID\n");
    }

    else if (count == 15 && (card_number_copy == 34 || card_number_copy == 37))
    {
        printf("AMEX\n");
    }

    else if (count == 16 && (card_number_copy >= 51 && card_number_copy <= 55))
    {
        printf("MASTERCARD\n");
    }

    else if ((count == 13 || count == 16) && card_number_copy / 10 == 4)
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}

File: projects/phase-02-computer-science-foundations/cs50x/week-06-python/sentimental-credit/credit.py
# A program that determines whether a provided credit card number is valid according to Luhn's algorithm and identifies the card type (AMEX, MASTERCARD, VISA) if valid.

while True:
    card_number = input("Number: ")
    if card_number.isdigit():
        break

checksum = 0

for digit in card_number[-2::-2]:
    product = int(digit) * 2
    checksum += (product % 10) + (product // 10)

for digit in card_number[-1::-2]:
    checksum += int(digit)

if checksum % 10 != 0:
    print("INVALID")

elif len(card_number) == 15 and card_number.startswith(("34", "37")):
    print("AMEX")

elif len(card_number) == 16 and card_number.startswith(("51", "52", "53", "54", "55")):
    print("MASTERCARD")

elif len(card_number) in [13, 16] and card_number.startswith(("4")):
    print("VISA")

else:
    print("INVALID")
